[Hack Week] Add Deselect action from long press

danielebogo opened the pull request:
Fixes PCIOS-397
Fixes PCIOS-271

This PR adds the bulk action to deselect all selected episodes above or below. So far we had the select all above or below but no deselect.

## To test

- Run this branch
- Make sure the Tracks event logger is on
- Open a Podcast detail
- Long press on an episode to enable the multi select option
- Long press again to show the sheet with Select all above and below
- Select one of the 2
- Then do it again
- Observe the deselect option is showed if necessary:
  - If you selected select all above, it will show deselect all above. Same for the below.
  - If all episodes are selected it will show both deselect all above and below
- Do the same test in:
  - Playlist detail
    - Make sure the deselect all above and below actions track `filterDeselectAllAbove` and `filterDeselectAllBelow`
  - UpNext (here the first long press plays the episode. You need to enable the select first)
  - Uploaded
  - Starred
  - Listening History
  - Downloads

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

danielebogo committed:
Add deselect all bulk actions

danielebogo committed:
Add new deselect all above and below for Playlist + analytics

danielebogo committed:
Add action to Podcast detail

danielebogo committed:
Use extension for reusability

danielebogo committed:
Use extension in Playlist

danielebogo committed:
Use extension in Podcast

danielebogo committed:
Use extension in UpNext

danielebogo committed:
Use extension in Uploaded

danielebogo committed:
Use extension in Starred

danielebogo committed:
Use extension in Listening history

danielebogo committed:
Use extension in Downloads

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

CookieyedCodes commented:
Yay for adding this to iOS finally, just some some notes when I got someone to implement it on android, so just to check the following @danielebogo 

1.selecting an episode would pop-up select all options, whist selecting one that's allready selected would bring up the deselect options,

2.  this would apply besides the first one due to keeping the exhisting select all above and below functionality for the first, 

3. also to copy android it would include the currently selected one in the selection as its from that point 😁

SergioEstevao commented:
I've tested and behaviour in general is working correctly, but I found two edge cases:
- Select all above, and Deselect All above shown when pressing on the first cell of the list
- Select all bellow and Deselect all bellow show when pressing on the last cell of the list

New tracks events are showing for playlists, but not for the other screens, is that expected?

danielebogo commented:
Thanks @SergioEstevao 

> New tracks events are showing for playlists, but not for the other screens, is that expected?

Yes, we don't have other events

> but I found two edge cases

Not really edge cases. It's how the prompt logic works even now in prod. We don't consider the first/last row. But it could be a good addition though.

SergioEstevao reviewed (approved):
Code looks, and behaviour is correct.

On a follow up PR we maybe can update the behaviour to the following edge cases:
- Hiding select above or bellow when tapping the first/last rows
- Not showing when there is only one element in the list?

danielebogo committed:
Fix logic considering sections

danielebogo commented:
@SergioEstevao I added the logic to consider the first and last rows. I also fixed a bug where it didn't select all rows in Listening history or Downloads as it didn't consider all the sections

SergioEstevao commented:
<!-- linear:isThreadRoot -->

I will retest them.

danielebogo committed:
Fix multiple selection extension

danielebogo committed:
Inject episode section

SergioEstevao reviewed (changes requested):
It's almost there majority of scenarios are working correctly we just have now one edge case:
- When selecting an episode on the podcast episode list, when sections for season are enable and you tap on the first episode you still see an Select all above option.

danielebogo committed:
Attempt to fix list with fake header

danielebogo committed:
Merge branch 'trunk' into danieleb/add-deselect-all-bulk-action

SergioEstevao reviewed (approved):
All working correctly now!
Code looks good!
:shipit:
